### ISSUE TITLE
Add single Authorization Token field to avoid copy pasting everytime in the requests | CROSS PROJECT ISSUE

### ISSUE THREAD

**Issue by sahil9001** [anitab-org/bridge-in-tech-backend#196]
**Is your feature request related to a problem? Please describe.**
Currently every request requires to enter the Authorisation Token every time, instead what we can do is add a single field for Authorisation so we don't have to enter it every time.

**Describe the solution you'd like**
Currently in every request we have a field like this,

<img width="654" alt="Screenshot 2021-02-18 at 10 43 04 AM" src="https://user-images.githubusercontent.com/32628578/108308824-72544f00-71d6-11eb-8d7a-9cda8586c241.png">

We can replace it with this, 

https://user-images.githubusercontent.com/32628578/108309136-01f9fd80-71d7-11eb-97f1-80c5c357861b.mov

Here we just have to login once and then we can access all the locked routes as shown in the video (We don't have to enter Authorization Token again and again)

This can be achieved by modifying the swagger docs settings in the `settings.py`.

**Describe alternatives you've considered**

**Additional context**

**Comment by sahil9001** [comment]
@mtreacy002  Can I start my first PR with this issue?

**Comment by sahil9001** [comment]
@mtreacy002 Please respond

**Comment by sahil9001** [comment]
@isabelcosta Is anyone here?

**Comment by mtreacy002** [comment]
@sahil9001 , please be patient in waiting for the response for any questions or issues. Although admins or maintainers will try their best in helping contributors, you should prepare up to 3-4 days as a rule of thumb for someone to respond. This could be due to they have limited bandwidth since they have other commitment or event in their life (e.g. work, family event, uni exam or assignment, etc), or simply because they have a different time zone to you (like when you posted this 15 hours ago, that was very late at night at my time and I was already in bed). 
Maintaining a respectful and positive interaction at all time is one of the key values of AnitaB Open Source community 😉.

**Comment by mtreacy002** [comment]
In regard to this PR, @sahil9001, can you please elaborate (with supporting articles) what is the pros and cons of having this approach (single authorisation token), like security issue vs efficiency, etc.

**Comment by mtreacy002** [comment]
In regard to this PR, @sahil9001, can you please elaborate (with supporting articles) what are the pros and cons of having this approach (single authorisation token), like security issue vs efficiency, etc.

**Comment by sahil9001** [comment]
> @sahil9001 , please be patient in waiting for the response for any questions or issues. Although admins or maintainers will try their best in helping contributors, you should prepare up to 3-4 days as a rule of thumb for someone to respond. This could be due to they have limited bandwidth since they have other commitment or event in their life (e.g. work, family event, uni exam or assignment, etc), or simply because they have a different time zone to you (like when you posted this 15 hours ago, that was very late at night at my time and I was already in bed).
> Maintaining a respectful and positive interaction at all time is one of the key values of AnitaB Open Source community 😉.

I'm sorry, I'll take care of this from next time.

**Comment by sahil9001** [comment]
> In regard to this PR, @sahil9001, can you please elaborate (with supporting articles) what are the pros and cons of having this approach (single authorisation token), like security issue vs efficiency, etc.

I didn't find any article regarding this, but I could support my point by few points:
 
- Single Auth Token is similar to the normal Auth Token we use, it just benefits the user by not entering the same token in every request again and again.

- It provides efficiency as it saves time, like we wouldn't want to find the Token, then copy it and then paste it in every request, this also creates redundant time for copy pasting it everytime in the request. Instead we could just put it once and authorize, then we could access any authorized route without entering the Token as we've already entered it.

- This is just a proposed feature which could increase efficiency by saving time.

**Comment by mtreacy002** [comment]
@sahil9001 , I'm concern this will raise security issue since we'll be keeping the token somewhere in our backend server instead of just utilising the jwt to do the authentication for us. We have to hear what @isabelcosta opinion on this as the Mentorship System is the one managing the user's authentication. Plus, if we do going that direction, this change needs to be done on the MS backend side first before BIT coz BIT will apply however MS decides to do its authentication for consistency across project.

**Comment by epicadk** [comment]
@sahil9001 Each request will still contain the authorization header right? This is just being done for the convenience of the user?

**Comment by sahil9001** [comment]
> @sahil9001 Each request will still contain the authorization header right? This is just being done for the convenience of the user?  

Yes, it'll still contain token in the header.

**Comment by epicadk** [comment]
> > @sahil9001 Each request will still contain the authorization header right? This is just being done for the convenience of the user?
> 
> Yes, it'll still contain token in the header.

Yep @mtreacy002 it makes sense. Instead of repeatedly copy pasting the same auth token everywhere we can just enter it once in swagger.

**Comment by sahil9001** [comment]
> > > @sahil9001 Each request will still contain the authorization header right? This is just being done for the convenience of the user?
> > 
> > Yes, it'll still contain token in the header.
> 
> Yep @mtreacy002 it makes sense. Instead of repeatedly copy pasting the same auth token everywhere we can just enter it once in swagger. 

Yeah that's what I'm saying here.😁

**Comment by mtreacy002** [comment]
@epicadk and @sahil9001 , you need to open an issue on MS backend first on this. As I explained, the JWT is within MS backend territory which BIT has to closely follows. Only once the cross project issue on MS backend has been resolved and merged into develop branch, BIT can continue working on this issue. I’ll label this issue a cross project issue and put status “on hold” for now.

**Comment by epicadk** [comment]
Well it isn't a change to the JWT per say. There will be **no change** in the api, only the swagger ui will be changed that is all. Will this also require change on the MS backend?

**Comment by mtreacy002** [comment]
Also, @sahil9001 and @epicadk , what about the security implication to this? It’s no longer the same as reliance on JWT instead what it does seems to cancel out the reason of why JWT is there in the first time (making sure web app is secure in regard to authentication). Can you please explain why you think this is much better option to JWT? Saving developer’s time is a lesser concern for me then security as in production, user doesn’t actually need to add anything (token) on the UI. But with user token expose somewhere in the backend server outside JWT protection, things can go wrong fast 🤔. @isabelcosta , can we have your thoughts on this please?

**Comment by mtreacy002** [comment]
> Will this also require change on the MS backend?

Yes, I believe so, @epicadk since we can’t have two different ways to interact with our 2 swagger UIs BridgeInTech is using for its backend server (one for BIT backend server and the other for the MS backend server). The developer will be confused, right?

**Comment by epicadk** [comment]
> > Will this also require change on the MS backend?
> 
> Yes, I believe so, @epicadk since we can’t have two different ways to interact with our 2 swagger UIs BridgeInTech is using for its backend server (one for BIT backend server and the other for the MS backend server). The developer will be confused, right?

Makes sense. We can open a similar issue on mentorship backend.

**Comment by sahil9001** [comment]
@mtreacy002 @epicadk Are we going to open issue in mentorship backend? I can't find this issue created on that repo.

**Comment by epicadk** [comment]
> @mtreacy002 @epicadk Are we going to open issue in mentorship backend? I can't find this issue created on that repo.

Nope hasn't been created. You can create the issue on the mentorship repo. @sahil9001

**Comment by sahil9001** [comment]
> > @mtreacy002 @epicadk Are we going to open issue in mentorship backend? I can't find this issue created on that repo.
> 
> Nope hasn't been created. You can create the issue on the mentorship repo. @sahil9001 

Okay sure! Will do by today!

**Comment by mtreacy002** [comment]
@sahil9001 , since the Mentorship System will stick with using jwt token for each api requests when user logged-in (see comment [here](https://github.com/anitab-org/mentorship-backend/issues/1014#issuecomment-787444760)), BIT will need to follow suit. So, I will also close this issue now.
cc @epicadk and @isabelcosta
